# Sources
https://www.kaggle.com/rounakbanik/movie-recommender-systems

## Simple Recommender

The Simple Recommender offers generalized recommnendations to every user based on movie popularity and (sometimes) genre. The basic idea behind this recommender is that movies that are more popular and more critically acclaimed will have a higher probability of being liked by the average audience. This model does not give personalized recommendations based on the user. 

The implementation of this model is extremely trivial. All we have to do is sort our movies based on ratings and popularity and display the top movies of our list. As an added step, we can pass in a genre argument to get the top movies of a particular genre. 

In [187]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from ast import literal_eval
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet

import warnings; warnings.simplefilter('ignore')

In [188]:
md = pd. read_csv('../data/movies_metadata.csv')
md.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview    ...     release_date  \
0  Led by Woody, Andy's toys live happily in his ...    ...       1995-10-30   
1  When siblings Judy and Peter discover an encha...    ...       1995-12-15   
2  A family wedding reignites the ancient feud be...    ...       1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...    ...       1995-12-22   
4  Just when George Banks has recovered from his ...    ...       1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [189]:
md.columns

Index([u'adult', u'belongs_to_collection', u'budget', u'genres', u'homepage',
       u'id', u'imdb_id', u'original_language', u'original_title', u'overview',
       u'popularity', u'poster_path', u'production_companies',
       u'production_countries', u'release_date', u'revenue', u'runtime',
       u'spoken_languages', u'status', u'tagline', u'title', u'video',
       u'vote_average', u'vote_count'],
      dtype='object')

In [190]:
md['genres'] = md['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

I use the TMDB Ratings to come up with our **Top Movies Chart.** I will use IMDB's *weighted rating* formula to construct my chart. Mathematically, it is represented as follows:

Weighted Rating (WR) = $(\frac{v}{v + m} . R) + (\frac{m}{v + m} . C)$

where,
* *v* is the number of votes for the movie
* *m* is the minimum votes required to be listed in the chart
* *R* is the average rating of the movie
* *C* is the mean vote across the whole report

The next step is to determine an appropriate value for *m*, the minimum votes required to be listed in the chart. We will use **95th percentile** as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 95% of the movies in the list.

I will build our overall Top 250 Chart and will define a function to build charts for a particular genre. Let's begin!

In [191]:
vote_counts = md[md['vote_count'].notnull()]['vote_count'].astype('int')
vote_averages = md[md['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_averages.mean()
C

5.244896612406511

In [192]:
m = vote_counts.quantile(0.95)
m

434.0

In [193]:
md['year'] = pd.to_datetime(md['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

In [194]:
qualified = md[(md['vote_count'] >= m) & (md['vote_count'].notnull()) & (md['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'genres','homepage','poster_path']]
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
qualified.shape

(2274, 8)

In [195]:
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [196]:
qualified['wr'] = qualified.apply(weighted_rating, axis=1)

In [197]:
qualified = qualified.sort_values('wr', ascending=False).head(250)

In [198]:
qualified.head(15)

title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
22879                                       Interstellar  2014       11187   
2843                                          Fight Club  1999        9678   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
292                                         Pulp Fiction  1994        8670   
314                             The Shawshank Redemption  1994        8358   
7000       The Lord of the Rings: The Return of the King  2003        8226   
351                                         Forrest Gump  1994        8147   
5814               The Lord of the Rings: The Two Towers  2002        7641   
256                                            Star Wars  1977        6778   
1225                                  Back to the Future  1985        6239   
834                                        The Godfather  1972        6024   
1154                             The Empire Strikes Back  1980        5998   
46                                                 Se7en  1995        5915   

       vote_average popularity  \
15480             8    29.1081   
12481             8    123.167   
22879             8    32.2135   
2843              8    63.8696   
4863              8    32.0707   
292               8     140.95   
314               8    51.6454   
7000              8    29.3244   
351               8    48.3072   
5814              8    29.4235   
256               8    42.1497   
1225              8    25.7785   
834               8    41.1093   
1154              8     19.471   
46                8    18.4574   

                                                  genres  \
15480  [Action, Thriller, Science Fiction, Mystery, A...   
12481                   [Drama, Action, Crime, Thriller]   
22879                [Adventure, Drama, Science Fiction]   
2843                                             [Drama]   
4863                        [Adventure, Fantasy, Action]   
292                                    [Thriller, Crime]   
314                                       [Drama, Crime]   
7000                        [Adventure, Fantasy, Action]   
351                             [Comedy, Drama, Romance]   
5814                        [Adventure, Fantasy, Action]   
256                 [Adventure, Action, Science Fiction]   
1225        [Adventure, Comedy, Science Fiction, Family]   
834                                       [Drama, Crime]   
1154                [Adventure, Action, Science Fiction]   
46                            [Crime, Mystery, Thriller]   

                                                homepage  \
15480              http://inceptionmovie.warnerbros.com/   
12481       http://thedarkknight.warnerbros.com/dvdsite/   
22879                  http://www.interstellarmovie.net/   
2843          http://www.foxmovies.com/movies/fight-club   
4863                      http://www.lordoftherings.net/   
292                                                  NaN   
314                                                  NaN   
7000                       http://www.lordoftherings.net   
351                                                  NaN   
5814                      http://www.lordoftherings.net/   
256    http://www.starwars.com/films/star-wars-episod...   
1225   http://www.backtothefuture.com/movies/backtoth...   
834                         http://www.thegodfather.com/   
1154   http://www.starwars.com/films/star-wars-episod...   
46                            http://www.sevenmovie.com/   

                            poster_path        wr  
15480  /qmDpIHrmpJINaRKAfWQfftjCdyi.jpg  7.917588  
12481  /1hRoyzDtpgMU7Dz4JF22RANzQO7.jpg  7.905871  
22879  /nBNZadXqJSdt05SHLqgT0HuC5Gm.jpg  7.897107  
2843   /adw6Lq9FiC9zjYEpOqfq03ituwp.jpg  7.881753  
4863   /bxVxZb5O9OxCO0oRUNdCnpy9NST.jpg  7.871787  
292     /dM2w364MScsjFf

In [199]:
s = md.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'genre'
gen_md = md.drop('genres', axis=1).join(s)

In [200]:
def build_chart(genre, percentile=0.85):
    df = gen_md[gen_md['genre'] == genre]
    vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(percentile)
    
    qualified = df[(df['vote_count'] >= m) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity','homepage','poster_path']]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    
    qualified['wr'] = qualified.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(250)
    
    return qualified

In [201]:
build_chart('Romance').head(15)

title  year  vote_count  vote_average popularity  \
10309  Dilwale Dulhania Le Jayenge  1995         661             9     34.457   
351                   Forrest Gump  1994        8147             8    48.3072   
876                        Vertigo  1958        1162             8    18.2082   
40251                   Your Name.  2016        1030             8  34.461252   
883               Some Like It Hot  1959         835             8    11.8451   
1132               Cinema Paradiso  1988         834             8     14.177   
19901                     Paperman  2012         734             8    7.19863   
37863                  Sing Street  2016         669             8  10.672862   
882                  The Apartment  1960         498             8    11.9943   
38718               The Handmaiden  2016         453             8  16.727405   
3189                   City Lights  1931         444             8    10.8915   
24886             The Way He Looks  2014         262             8    5.71127   
45437               In a Heartbeat  2017         146             8   20.82178   
1639                       Titanic  1997        7770             7    26.8891   
19731      Silver Linings Playbook  2012        4840             7    14.4881   

                                                homepage  \
10309                                                NaN   
351                                                  NaN   
876                                                  NaN   
40251    https://www.funimationfilms.com/movie/yourname/   
883                                                  NaN   
1132                                                 NaN   
19901  http://www.disneyanimation.com/projects/shorts...   
37863                                                NaN   
882                                                  NaN   
38718                                                NaN   
3189   http://www.charliechaplin.com/biography/articl...   
24886                                                NaN   
45437                                                NaN   
1639                         http://www.titanicmovie.com   
19731             http://silverliningsplaybookmovie.com/   

                            poster_path        wr  
10309  /2gvbZMtV1Zsl7FedJa5ysbpBx2G.jpg  8.565285  
351    /yE5d3BUhE8hCnkMUJOo1QDoOGNz.jpg  7.971357  
876    /obhM86qyv8RsE69XSMTtT9FdE0b.jpg  7.811667  
40251  /xq1Ugd62d23K2knRUx6xxuALTZB.jpg  7.789489  
883    /pxc9EFCMYkItESpqqrI783yl8Gh.jpg  7.745154  
1132   /xKAweeF2ZPMNn6ce4GclSbr59Pv.jpg  7.744878  
19901  /3TpMBcAYH4cxCw5WoRacWodMTCG.jpg  7.713951  
37863  /s0C78plmx3dFcO3WMnoXCz56FiN.jpg  7.689483  
882    /6OH3HlplEGoGjqeqRmWaVBGqbav.jpg  7.599317  
38718  /wvzfK5QR6dGLwND8MCzWjsQWG4Q.jpg  7.566166  
3189   /bXNvzjULc9jrOVhGfjcc64uKZmZ.jpg  7.558867  
24886  /oAPCsQiWV6YUd0Gt62BOwb8aSth.jpg  7.331363  
45437  /wJUJROdLOtOzMixkjkx1aaZGSLl.jpg  7.003959  
1639   /kHXEpyfl6zqn8a6YuozZUujufXf.jpg  6.981546  
19731  /ilrZAV2klTB0FLxLb01bOp5pzD9.jpg  6.970581

In [219]:
genreSet = list(set(s))
dfGenre = build_chart(genreSet[2]).head(5)
dfGenre['Genre'] = genreSet[2]
for i in range(3,len(genreSet)):
    try:
        x = build_chart(genreSet[i]).head(5)
        x['Genre'] = genreSet[i]
        dfGenre = dfGenre.append(x)
    except:
        continue

In [220]:
dfGenre

title  year  vote_count  \
10309                        Dilwale Dulhania Le Jayenge  1995         661   
351                                         Forrest Gump  1994        8147   
876                                              Vertigo  1958        1162   
40251                                         Your Name.  2016        1030   
883                                     Some Like It Hot  1959         835   
359                                        The Lion King  1994        5520   
5481                                       Spirited Away  2001        3968   
9698                                Howl's Moving Castle  2004        2049   
2884                                   Princess Mononoke  1997        2041   
5833                                  My Neighbor Totoro  1988        1730   
37863                                        Sing Street  2016         669   
21556                          One Direction: This Is Us  2013         209   
40882                                         La La Land  2016        4745   
588                                 Beauty and the Beast  1991        3029   
19384                                      Pitch Perfect  2012        2310   
10309                        Dilwale Dulhania Le Jayenge  1995         661   
351                                         Forrest Gump  1994        8147   
1225                                  Back to the Future  1985        6239   
18465                                   The Intouchables  2011        5410   
22841                           The Grand Budapest Hotel  2014        4644   
24860                                 The Imitation Game  2014        5895   
522                                     Schindler's List  1993        4436   
1165                                      Apocalypse Now  1979        2112   
5857                                         The Pianist  2002        1927   
732    Dr. Strangelove or: How I Learned to Stop Worr...  1964        1472   
1213                                         The Shining  1980        3890   
1176                                              Psycho  1960        2405   
1171                                               Alien  1979        4564   
41492                                              Split  2016        4461   
14236                                         Zombieland  2009        3655   
...                                                  ...   ...         ...   
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
7000       The Lord of the Rings: The Return of the King  2003        8226   
5814               The Lord of the Rings: The Two Towers  2002        7641   
43320                                    Planet Earth II  2016          50   
16881                                              Senna  2010         282   
21556                          One Direction: This Is Us  2013         209   
39085                                       Planet Earth  2006         176   
23325                                  Jodorowsky's Dune  2013         172   
24860                                 The Imitation Game  2014        5895   
522                                     Schindler's List  1993        4436   
38244                                           11.22.63  2016         213   
6669                                        Come and See  1985         174   
6833                          The Passion of Joan of Arc  1928         159   
1225                                  Back to the Future  1985        6239   
359                                        The Lion King  1994        5520   
5481                                       Spirited Away  2001        3968   
5833                                  My Neighbor Totoro  1988        1730   
926                                It's a Wonderful Life  1946        110

In [222]:
dfGenre.homepage.fillna('https://www.google.com/search?q=' + dfGenre.title.str.replace(' ','+'), inplace=True)

In [223]:
dfGenre

title  year  vote_count  \
10309                        Dilwale Dulhania Le Jayenge  1995         661   
351                                         Forrest Gump  1994        8147   
876                                              Vertigo  1958        1162   
40251                                         Your Name.  2016        1030   
883                                     Some Like It Hot  1959         835   
359                                        The Lion King  1994        5520   
5481                                       Spirited Away  2001        3968   
9698                                Howl's Moving Castle  2004        2049   
2884                                   Princess Mononoke  1997        2041   
5833                                  My Neighbor Totoro  1988        1730   
37863                                        Sing Street  2016         669   
21556                          One Direction: This Is Us  2013         209   
40882                                         La La Land  2016        4745   
588                                 Beauty and the Beast  1991        3029   
19384                                      Pitch Perfect  2012        2310   
10309                        Dilwale Dulhania Le Jayenge  1995         661   
351                                         Forrest Gump  1994        8147   
1225                                  Back to the Future  1985        6239   
18465                                   The Intouchables  2011        5410   
22841                           The Grand Budapest Hotel  2014        4644   
24860                                 The Imitation Game  2014        5895   
522                                     Schindler's List  1993        4436   
1165                                      Apocalypse Now  1979        2112   
5857                                         The Pianist  2002        1927   
732    Dr. Strangelove or: How I Learned to Stop Worr...  1964        1472   
1213                                         The Shining  1980        3890   
1176                                              Psycho  1960        2405   
1171                                               Alien  1979        4564   
41492                                              Split  2016        4461   
14236                                         Zombieland  2009        3655   
...                                                  ...   ...         ...   
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
7000       The Lord of the Rings: The Return of the King  2003        8226   
5814               The Lord of the Rings: The Two Towers  2002        7641   
43320                                    Planet Earth II  2016          50   
16881                                              Senna  2010         282   
21556                          One Direction: This Is Us  2013         209   
39085                                       Planet Earth  2006         176   
23325                                  Jodorowsky's Dune  2013         172   
24860                                 The Imitation Game  2014        5895   
522                                     Schindler's List  1993        4436   
38244                                           11.22.63  2016         213   
6669                                        Come and See  1985         174   
6833                          The Passion of Joan of Arc  1928         159   
1225                                  Back to the Future  1985        6239   
359                                        The Lion King  1994        5520   
5481                                       Spirited Away  2001        3968   
5833                                  My Neighbor Totoro  1988        1730   
926                                It's a Wonderful Life  1946        110

In [225]:
dfGenre.to_csv("../data/top5Genre.csv")

## Content Based Recommender

The recommender we built in the previous section suffers some severe limitations. For one, it gives the same recommendation to everyone, regardless of the user's personal taste. If a person who loves romantic movies (and hates action) were to look at our Top 15 Chart, s/he wouldn't probably like most of the movies. If s/he were to go one step further and look at our charts by genre, s/he wouldn't still be getting the best recommendations.

For instance, consider a person who loves *Dilwale Dulhania Le Jayenge*, *My Name is Khan* and *Kabhi Khushi Kabhi Gham*. One inference we can obtain is that the person loves the actor Shahrukh Khan and the director Karan Johar. Even if s/he were to access the romance chart, s/he wouldn't find these as the top recommendations.

To personalise our recommendations more, I am going to build an engine that computes similarity between movies based on certain metrics and suggests movies that are most similar to a particular movie that a user liked. Since we will be using movie metadata (or content) to build this engine, this also known as **Content Based Filtering.**

I will build two Content Based Recommenders based on:
* Movie Overviews and Taglines
* Movie Cast, Crew, Keywords and Genre

Also, as mentioned in the introduction, I will be using a subset of all the movies available to us due to limiting computing power available to me. 

In [227]:
links_small = pd.read_csv('../data/links_small.csv')
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')

In [228]:
md = md.drop([19730, 29503, 35587])

In [229]:
#Check EDA Notebook for how and why I got these indices.
md['id'] = md['id'].astype('int')

In [230]:
smd = md[md['id'].isin(links_small)]
smd.shape

(9099, 25)

We have **9099** movies avaiable in our small movies metadata dataset which is 5 times smaller than our original dataset of 45000 movies.

### Movie Description Based Recommender

Let us first try to build a recommender using movie descriptions and taglines. We do not have a quantitative metric to judge our machine's performance so this will have to be done qualitatively.

In [232]:
smd['tagline'] = smd['tagline'].fillna('')
smd['description'] = smd['overview'] + smd['tagline']
smd['description'] = smd['description'].fillna('')

In [233]:
tf = TfidfVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
tfidf_matrix = tf.fit_transform(smd['description'])

In [234]:
tfidf_matrix.shape

(9099, 268124)

#### Cosine Similarity

I will be using the Cosine Similarity to calculate a numeric quantity that denotes the similarity between two movies. Mathematically, it is defined as follows:

$cosine(x,y) = \frac{x. y^\intercal}{||x||.||y||} $

Since we have used the TF-IDF Vectorizer, calculating the Dot Product will directly give us the Cosine Similarity Score. Therefore, we will use sklearn's **linear_kernel** instead of cosine_similarities since it is much faster.

In [236]:
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

In [237]:
cosine_sim[0]

array([ 1.        ,  0.00680476,  0.        , ...,  0.        ,
        0.00344913,  0.        ])

We now have a pairwise cosine similarity matrix for all the movies in our dataset. The next step is to write a function that returns the 30 most similar movies based on the cosine similarity score.

In [239]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

In [240]:
def get_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:31]
    movie_indices = [i[0] for i in sim_scores]
    return titles.iloc[movie_indices]

We're all set. Let us now try and get the top recommendations for a few movies and see how good the recommendations are.

In [242]:
get_recommendations('The Godfather').head(10)

973      The Godfather: Part II
8387                 The Family
3509                       Made
4196         Johnny Dangerously
29               Shanghai Triad
5667                       Fury
2412             American Movie
1582    The Godfather: Part III
4221                    8 Women
2159              Summer of Sam
Name: title, dtype: object

We see that for **The Dark Knight**, our system is able to identify it as a Batman film and subsequently recommend other Batman films as its top recommendations. But unfortunately, that is all this system can do at the moment. This is not of much use to most people as it doesn't take into considerations very important features such as cast, crew, director and genre, which determine the rating and the popularity of a movie. Someone who liked **The Dark Knight** probably likes it more because of Nolan and would hate **Batman Forever** and every other substandard movie in the Batman Franchise.

Therefore, we are going to use much more suggestive metadata than **Overview** and **Tagline**. In the next subsection, we will build a more sophisticated recommender that takes **genre**, **keywords**, **cast** and **crew** into consideration.

### Metadata Based Recommender

To build our standard metadata based content recommender, we will need to merge our current dataset with the crew and the keyword datasets. Let us prepare this data as our first step.

In [243]:
credits = pd.read_csv('../data/credits.csv')
keywords = pd.read_csv('../data/keywords.csv')

In [244]:
keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')
md['id'] = md['id'].astype('int')

In [245]:
md.shape

(45463, 25)

In [246]:
md = md.merge(credits, on='id')
md = md.merge(keywords, on='id')

In [247]:
smd = md[md['id'].isin(links_small)]
smd.shape

(9219, 28)

We now have our cast, crew, genres and credits, all in one dataframe. Let us wrangle this a little more using the following intuitions:

1. **Crew:** From the crew, we will only pick the director as our feature since the others don't contribute that much to the *feel* of the movie.
2. **Cast:** Choosing Cast is a little more tricky. Lesser known actors and minor roles do not really affect people's opinion of a movie. Therefore, we must only select the major characters and their respective actors. Arbitrarily we will choose the top 3 actors that appear in the credits list. 

In [248]:
smd['cast'] = smd['cast'].apply(literal_eval)
smd['crew'] = smd['crew'].apply(literal_eval)
smd['keywords'] = smd['keywords'].apply(literal_eval)
smd['cast_size'] = smd['cast'].apply(lambda x: len(x))
smd['crew_size'] = smd['crew'].apply(lambda x: len(x))

In [249]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

In [250]:
smd['director'] = smd['crew'].apply(get_director)

In [251]:
smd['cast'] = smd['cast'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
smd['cast'] = smd['cast'].apply(lambda x: x[:3] if len(x) >=3 else x)

In [252]:
smd['keywords'] = smd['keywords'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

My approach to building the recommender is going to be extremely *hacky*. What I plan on doing is creating a metadata dump for every movie which consists of **genres, director, main actors and keywords.** I then use a **Count Vectorizer** to create our count matrix as we did in the Description Recommender. The remaining steps are similar to what we did earlier: we calculate the cosine similarities and return movies that are most similar.

These are steps I follow in the preparation of my genres and credits data:
1. **Strip Spaces and Convert to Lowercase** from all our features. This way, our engine will not confuse between **Johnny Depp** and **Johnny Galecki.** 
2. **Mention Director 3 times** to give it more weight relative to the entire cast.

In [253]:
smd['cast'] = smd['cast'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [254]:
smd['director'] = smd['director'].astype('str').apply(lambda x: str.lower(x.replace(" ", "")))
smd['director'] = smd['director'].apply(lambda x: [x,x, x])

#### Keywords

We will do a small amount of pre-processing of our keywords before putting them to any use. As a first step, we calculate the frequenct counts of every keyword that appears in the dataset.

In [255]:
s = smd.apply(lambda x: pd.Series(x['keywords']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'keyword'

In [256]:
s = s.value_counts()
s[:5]

independent film        610
woman director          550
murder                  399
duringcreditsstinger    327
based on novel          318
Name: keyword, dtype: int64

Keywords occur in frequencies ranging from 1 to 610. We do not have any use for keywords that occur only once. Therefore, these can be safely removed. Finally, we will convert every word to its stem so that words such as *Dogs* and *Dog* are considered the same.

In [257]:
s = s[s > 1]

In [258]:
stemmer = SnowballStemmer('english')
stemmer.stem('dogs')

u'dog'

In [259]:
def filter_keywords(x):
    words = []
    for i in x:
        if i in s:
            words.append(i)
    return words

In [267]:
import sys
reload(sys)
sys.setdefaultencoding("ISO-8859-1")
smd['keywords'] = smd['keywords'].apply(filter_keywords)
smd['keywords'] = smd['keywords'].apply(lambda x: [stemmer.stem(i) for i in x])
smd['keywords'] = smd['keywords'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

TypeError: descriptor 'lower' requires a 'str' object but received a 'unicode'

In [268]:
smd['soup'] = smd['keywords'] + smd['cast'] + smd['director'] + smd['genres']
smd['soup'] = smd['soup'].apply(lambda x: ' '.join(x))

In [269]:
count = CountVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
count_matrix = count.fit_transform(smd['soup'])

In [270]:
cosine_sim = cosine_similarity(count_matrix, count_matrix)

In [271]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

In [272]:
get_recommendations('The Dark Knight').head(10)

6218                         Batman Begins
8031                 The Dark Knight Rises
1134                        Batman Returns
6623                          The Prestige
7659            Batman: Under the Red Hood
2085                             Following
1260                        Batman & Robin
4145                              Insomnia
7648                             Inception
9024    Batman v Superman: Dawn of Justice
Name: title, dtype: object

In [ ]:
idx = indices[title]
sim_scores = list(enumerate(cosine_sim[idx]))
sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
sim_scores = sim_scores[1:31]
movie_indices = [i[0] for i in sim_scores]
#return titles.iloc[movie_indices]

In [289]:
titleDF = pd.DataFrame(titles)


In [291]:
cosine_simDF = pd.DataFrame(cosine_sim)

In [293]:
cosine_simDF

0         1         2         3         4         5         6     \
0     1.000000  0.037980  0.045326  0.019854  0.018990  0.000000  0.020332   
1     0.037980  1.000000  0.000000  0.022244  0.000000  0.000000  0.000000   
2     0.045326  0.000000  1.000000  0.053093  0.025392  0.000000  0.054373   
3     0.019854  0.022244  0.053093  1.000000  0.044488  0.015813  0.071449   
4     0.018990  0.000000  0.025392  0.044488  1.000000  0.000000  0.045561   
5     0.000000  0.000000  0.000000  0.015813  0.000000  1.000000  0.000000   
6     0.020332  0.000000  0.054373  0.071449  0.045561  0.000000  1.000000   
7     0.025055  0.056143  0.000000  0.029348  0.000000  0.039912  0.000000   
8     0.000000  0.023980  0.000000  0.000000  0.000000  0.034095  0.000000   
9     0.000000  0.016011  0.000000  0.000000  0.000000  0.022764  0.000000   
10    0.040664  0.022780  0.054373  0.119081  0.022780  0.016194  0.048780   
11    0.022663  0.000000  0.030303  0.026547  0.025392  0.000000  0.027186   
12    0.042806  0.047960  0.000000  0.000000  0.000000  0.000000  0.000000   
13    0.000000  0.000000  0.000000  0.023256  0.000000  0.015813  0.000000   
14    0.000000  0.023980  0.000000  0.000000  0.000000  0.017047  0.025675   
15    0.000000  0.000000  0.000000  0.025071  0.000000  0.068190  0.000000   
16    0.017244  0.019320  0.023057  0.141393  0.000000  0.027470  0.020686   
17    0.034488  0.019320  0.023057  0.020199  0.019320  0.041204  0.020686   
18    0.021403  0.023980  0.028618  0.050141  0.047960  0.034095  0.051350   
19    0.051732  0.038641  0.023057  0.040398  0.038641  0.054939  0.165486   
20    0.016402  0.018377  0.021932  0.076852  0.018377  0.104515  0.019676   
21    0.000000  0.000000  0.000000  0.021354  0.000000  0.043561  0.000000   
22    0.014647  0.016411  0.000000  0.000000  0.000000  0.070000  0.000000   
23    0.031346  0.017560  0.000000  0.036717  0.017560  0.024967  0.018801   
24    0.000000  0.000000  0.020101  0.052827  0.000000  0.023947  0.018033   
25    0.048351  0.000000  0.032325  0.028318  0.000000  0.019256  0.000000   
26    0.059561  0.022244  0.026547  0.069767  0.044488  0.015813  0.023816   
27    0.000000  0.000000  0.029424  0.077331  0.000000  0.017528  0.026398   
28    0.015905  0.053461  0.000000  0.000000  0.017820  0.000000  0.019080   
29    0.000000  0.000000  0.000000  0.024419  0.000000  0.049814  0.000000   
...        ...       ...       ...       ...       ...       ...       ...   
9189  0.000000  0.000000  0.000000  0.000000  0.000000  0.018051  0.000000   
9190  0.015237  0.017072  0.020374  0.017849  0.017072  0.072820  0.018279   
9191  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
9192  0.018990  0.063830  0.025392  0.066733  0.042553  0.015125  0.182242   
9193  0.019407  0.000000  0.025950  0.022733  0.021744  0.000000  0.023281   
9194  0.000000  0.000000  0.000000  0.065357  0.000000  0.000000  0.000000   
9195  0.000000  0.000000  0.000000  0.022733  0.000000  0.030916  0.000000   
9196  0.045326  0.025392  0.030303  0.026547  0.025392  0.018051  0.027186   
9197  0.021403  0.000000  0.028618  0.025071  0.023980  0.017047  0.025675   
9198  0.000000  0.000000  0.000000  0.000000  0.000000  0.038511  0.000000   
9199  0.000000  0.000000  0.000000  0.000000  0.000000  0.022628  0.000000   
9200  0.000000  0.023357  0.000000  0.000000  0.000000  0.016605  0.000000   
9201  0.000000  0.000000  0.000000  0.000000  0.000000  0.038511  0.000000   
9202  0.025055  0.000000  0.033501  0.029348  0.028072  0.000000  0.030056   
9203  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
9204  0.018230  0.040850  0.048751  0.106770  0.040850  0.029040  0.043737   
9205  0.054293  0.030415  0.000000  0.000000  0.000000  0.000000  0.000000   
9206  0.016949  0.075960  0.000000  0.099268  0.000000  0.027000  0.000000   
9207  0.000000  0.000000  0.000000  0.000000  0.000000  0.043244  0.000000   
9208  0.000000  0.000000  0.0000

In [296]:
totalDF = pd.concat([titleDF, cosine_simDF],axis =1)

In [297]:
totalDF

title         0         1  \
0                                             Toy Story  1.000000  0.037980   
1                                               Jumanji  0.037980  1.000000   
2                                      Grumpier Old Men  0.045326  0.000000   
3                                     Waiting to Exhale  0.019854  0.022244   
4                           Father of the Bride Part II  0.018990  0.000000   
5                                                  Heat  0.000000  0.000000   
6                                               Sabrina  0.020332  0.000000   
7                                          Tom and Huck  0.025055  0.056143   
8                                          Sudden Death  0.000000  0.023980   
9                                             GoldenEye  0.000000  0.016011   
10                               The American President  0.040664  0.022780   
11                          Dracula: Dead and Loving It  0.022663  0.000000   
12                                                Balto  0.042806  0.047960   
13                                                Nixon  0.000000  0.000000   
14                                     Cutthroat Island  0.000000  0.023980   
15                                               Casino  0.000000  0.000000   
16                                Sense and Sensibility  0.017244  0.019320   
17                                           Four Rooms  0.034488  0.019320   
18                       Ace Ventura: When Nature Calls  0.021403  0.023980   
19                                          Money Train  0.051732  0.038641   
20                                           Get Shorty  0.016402  0.018377   
21                                              Copycat  0.000000  0.000000   
22                                            Assassins  0.014647  0.016411   
23                                               Powder  0.031346  0.017560   
24                                    Leaving Las Vegas  0.000000  0.000000   
25                                              Othello  0.048351  0.000000   
26                                         Now and Then  0.059561  0.022244   
27                                           Persuasion  0.000000  0.000000   
28                            The City of Lost Children  0.015905  0.053461   
29                                       Shanghai Triad  0.000000  0.000000   
...                                                 ...       ...       ...   
9189                                       Kill Command  0.000000  0.000000   
9190                                   Now You See Me 2  0.015237  0.017072   
9191                                     The Video Dead  0.000000  0.000000   
9192   Teenage Mutant Ninja Turtles: Out of the Shadows  0.018990  0.063830   
9193                 Popstar: Never Stop Never Stopping  0.019407  0.000000   
9194                                    The Conjuring 2  0.000000  0.000000   
9195                            Approaching the Unknown  0.000000  0.000000   
9196                                       Ghostbusters  0.045326  0.025392   
9197                               Central Intelligence  0.021403  0.000000   
9198                                       Jason Bourne  0.000000  0.000000   
9199                                    The Maid's Room  0.000000  0.000000   
9200                               The Legend of Tarzan  0.000000  0.023357   
9201                           The Purge: Election Year  0.000000  0.000000   
9202                   Mike and Dave Need Wedding Dates  0.025055  0.000000   
9203                                Survive and Advance  0.000000  0.000000   
9204                                           Tallulah  0.018230  0.040850   
9205                                              Piper  0.054293  0.030415   
9206                                              Nerve  0.016949  0.075960   
9207                              My Friend Rockefeller  0.000000  0.000000   
9208                 

In [298]:
totalDF.to_csv('../data/contentRec.csv')